akka-remote/src/main/scala/akka/remote/artery/ArteryTcpTransport.scala

```diff
@@ -0,0 +1,352 @@
+        .addAttributes(Attributes.logLevels(onElement = LogLevels.Off, onFailure = Logging.WarningLevel))
+
+      // FIXME config of backoff
+      import scala.concurrent.duration._
+      if (streamId == ControlStreamId) {
+        // restart of inner connection part important in control flow, since system messages
+        // are buffered and resent from the outer SystemMessageDelivery stage.
+        // FIXME The mat value Future is currently never completed, because RestartFlow will retry forever, should it give up?
+        //       related to give-up-system-message-after ?
+        RestartFlow.withBackoff[ByteString, ByteString](1.second, 5.seconds, 0.1)(flowFactory)
+      } else {
+        // Best effort retry a few times
+        // FIXME only restart on failures?, but missing in RestartFlow, see https://github.com/akka/akka/pull/23911
+        RestartFlow.withBackoff[ByteString, ByteString](1.second, 5.seconds, 0.1, maxRestarts = 3)(flowFactory)
+      }
+    }
+
+    Flow[EnvelopeBuffer]
+      .map { env ⇒
+        val bytes = ByteString(env.byteBuffer)
+        bufferPool.release(env)
+        bytes
+      }
+      .recoverWithRetries(1, { case ArteryTransport.ShutdownSignal ⇒ Source.empty })
+      .via(connectionFlowWithRestart)
+      .map(_ ⇒ throw new IllegalStateException(s"Unexpected incoming bytes in outbound connection to [${outboundContext.remoteAddress}]"))
+      .toMat(Sink.ignore)(Keep.right)
+  }
+
+  override protected def runInboundStreams(): Unit = {
+    implicit val mat = materializer
+    implicit val sys = system
+
+    val (controlStream, controlStreamCompleted) = runInboundControlStream()
+    val (ordinaryMessagesStream, ordinaryMessagesStreamCompleted) = runInboundOrdinaryMessagesStream()
+    val (largeMessagesStream, largeMessagesStreamCompleted) = {
+      if (largeMessageChannelEnabled)
+        runInboundLargeMessagesStream()
+      else
+        (
+          Flow[EnvelopeBuffer]
+          .map(_ ⇒ log.warning("Dropping large message, missing large-message-destinations configuration."))
+          .to(Sink.ignore),
+          Promise[Done]().future) // never completed, not enabled
+    }
+
+    val inboundStream: Sink[EnvelopeBuffer, NotUsed] =
+      Sink.fromGraph(GraphDSL.create() { implicit b ⇒
+        import GraphDSL.Implicits._
+        val partition = b.add(Partition[EnvelopeBuffer](3, env ⇒ {
+          val streamId: Int = env.byteBuffer.get(EnvelopeBuffer.StreamIdOffset).toInt
+          if (streamId == OrdinaryStreamId) 1
+          else if (streamId == ControlStreamId) 0
+          else if (streamId == LargeStreamId) 2
+          else throw new IllegalArgumentException(s"Unexpected streamId [$streamId]")
+        }))
+        partition.out(0) ~> controlStream
+        partition.out(1) ~> ordinaryMessagesStream
+        partition.out(2) ~> largeMessagesStream
+        SinkShape(partition.in)
```

Comment:
Is the diagram out of date here or am i getting confused? i.e. there's no merge hub?

Reply:
Oh i found them

Reply:
Do we need the merge hub as we direct all connections into the same stream?

Reply:
That is an interesting observation. I have tried to keep as much as possible the same as the design for the Aeron streams, and there we can only have one single inbound since everything comes in via the single AeronSource.

For TCP we could materialize the inbound streams for each inbound connection, i.e. running many completely separate inbound streams. It would still have to include all 3 control, ordinary, large parts for each connection even though only one is used by a specific connection. Unless we can dynamically choose what to materialize based on the `streamId` (as a first byte, or in first frame of the connection), which is complicated.

However, this would be make the design for Aeron and TCP more different and more things might have to be changed, such as compression advertisments and materialized values. Number of inbound streams would be dynamic, and so on.

I'll add this as a design comment in the code.

Reply:
And I don't know what would be most scalable for many inbound connections. Would the MergeHub become the bottleneck or would running many larger streams become wasteful?

Reply:
One crazy idea, as perhaps a small improvement, could be to use `recoverWith` instead of `Partition`. First attach it to the `ordinaryMessagesStream` and when a non matching streamId arrives throw and recoverWith `controlStream` and so on. Adding that as a TODO for investigation.

Reply:
I think keeping them as similar as possible for now is the best option